### PR TITLE
feat: porting wemix

### PR DESCRIPTION
**Pull request by egonspace**
This enables Wemix PoA consensus engine so the node can sync Wemix legacy blocks.

defined `wemix consensus engine`
- it has two sub engines: wpoa, wbft
- wpoa: wemix legacy engine; only for sync, no mining
- wbft: currently qbft engine; it fails to mine a wemix block currently; in next PR, we should fix qbft engine for wemix block
- it switches engine at `MontBlanc` hard fork

How to test:
- init chain with `geth init --datadir ~/.ethereum/devnet ~/.ethereum/devnet/genesis.json` with following genesis.json file
- start geth with `geth --datadir ~/.ethereum/devnet --nodiscover --syncmode="full" --config ~/.ethereum/mainnet/config.toml`

```
{
  "alloc": {
    "0x329de77Fe7704C0aD96575674EA4CD5E81f3a899": {
      "balance": "0x7c13bc4b2c133c56000000"
    },
    "0x4675153A8b0B2C20CebCec1E7c54df6F69e64347": {
      "balance": "0x7c13bc4b2c133c56000000"
    },
    "0x6859Efc328Fa481D1a1f3f8cD5b33FAfC1bA5629": {
      "balance": "0xaf298d050e4395d69670b12b7f41000000000000"
    },
    "0x85731E7727850aaB592EFb119b4e363eAED10c7C": {
      "balance": "0x7c13bc4b2c133c56000000"
    },
    "0x97fe70530e77d007407e89D7b469f6359F57BA1A": {
      "balance": "0x7c13bc4b2c133c56000000"
    },
    "0xD88498767Fc2aA635D463110c4D9198054969F6b": {
      "balance": "0x7c13bc4b2c133c56000000"
    }
  },
  "coinbase": "0x6859Efc328Fa481D1a1f3f8cD5b33FAfC1bA5629",
  "config": {
    "applepieBlock": 0,
    "berlinBlock": 0,
    "byzantiumBlock": 0,
    "chainId": 1113,
    "constantinopleBlock": 0,
    "daoForkBlock": 0,
    "daoForkSupport": true,
    "eip150Block": 0,
    "eip155Block": 0,
    "eip158Block": 0,
    "homesteadBlock": 0,
    "istanbulBlock": 0,
    "londonBlock": 0,
    "muirGlacierBlock": 0,
    "petersburgBlock": 0,
    "pangyoBlock": 0,
    "briocheBlock": 607147,
    "montblancBlock": 10000000,
    "brioche": {
      "blockReward": 1000000000000000000,
      "firstHalvingBlock": 607147,
      "halvingPeriod": 63115200,
      "finishRewardBlock": 2412358474,
      "halvingTimes": 16,
      "halvingRate": 50
    },
    "qbft": {
      "epochLength": 10,
      "blockPeriodSeconds": 3,
      "emptyBlockPeriodSeconds": 0,
      "requestTimeoutSeconds": 1000,
      "proposerPolicy": 0,
      "blockReward": 1000000000000000000,
      "beneficiaryMode": "validator",
      "miningBeneficiary": "0x0000000000000000000000000000000000000000",
      "validatorSelectionMode": "blockheader",
      "validators": ["0x8455f21eCC18579e1Ba85e9F3d32730D6C328Fb0"],
      "maxRequestTimeoutSeconds": 4
    }
  },
  "difficulty": "0x1",
  "extraData": "0x54686520626567696e6e696e67206f662057656d6978332e3020746573746e6574206f6e204a756c79203173742c20323032320a30786637633835353333366339636565323835656334383963613730373765336361646363376635353365333861376566636434366137656332623538613565323935613839383438663336666639626335303963663736343439356564393264396135303730353061656131663862346265323831383765306333303234393130",
  "gasLimit": "105000000",
  "minerNodeId": "0x0",
  "minerNodeSig": "0x0",
  "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000042",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "rewards": "0x",
  "timestamp": "0x00"
}
```

### CODE REVIEW

File: accounts/abi/bind/bind_test.go [hunk 1]
```diff
@@ -297,7 +297,7 @@ var bindTests = []struct {
 			key, _ := crypto.GenerateKey()
 			auth, _ := bind.NewKeyedTransactorWithChainID(key, big.NewInt(1337))
 
-			sim := backends.NewSimulatedBackend(types.GenesisAlloc{auth.From: {Balance: big.NewInt(10000000000000000)}}, 10000000)
+			sim := backends.NewSimulatedBackend(types.GenesisAlloc{auth.From: {Balance: big.NewInt(1000000000000000000)}}, 10000000)
```
Comment (egonspace) [inline accounts/abi/bind/bind_test.go:300]: Wemix PoA suggests 100 gwei for gas tip cap that was 1 gwei in ethereum, so test account should have  more base coin.

File: core/genesis.go [hunk 2]
```diff
@@ -486,6 +491,24 @@ func (g *Genesis) MustCommit(db ethdb.Database, triedb *triedb.Database) *types.
...
+		panic("Cannot parse default wemix mainnet genesis.")
+	}
+	genesis.Config = params.WemixMainnetChainConfig
+	return genesis
+}
+
+func DefaultWemixTestnetGenesisBlock() *Genesis {
+	genesis := new(Genesis)
+	if err := json.NewDecoder(strings.NewReader(wemixTestnetGenesisJson)).Decode(genesis); err != nil {
+		panic("Cannot parse default wemix mainnet genesis.")
```
Comment (jed-wemade) [inline core/genesis.go]: mainnet -> testnet

File: core/rawdb/accessors_metadata.go [hunk 3]
```diff
@@ -162,7 +162,7 @@ func UpdateUncleanShutdownMarker(db ethdb.KeyValueStore) {
 	} else if err := rlp.DecodeBytes(data, &uncleanShutdowns); err != nil {
 		log.Warn("Error decoding unclean shutdown markers", "error", err)
 	}
-	// This shouldn't happen because we push a marker on Backend instantiation
+	// This shouldn't happen because we push a marker on backend instantiation
```
Comment (jed-wemade) [inline core/rawdb/accessors_metadata.go]: I think this is an unintended change.

File: core/types/block.go [hunk 4]
```diff
@@ -99,18 +105,19 @@ type Header struct {
...
-	Difficulty    *hexutil.Big
-	Number        *hexutil.Big
-	GasLimit      hexutil.Uint64
-	GasUsed       hexutil.Uint64
-	Time          hexutil.Uint64
-	Extra         hexutil.Bytes
-	BaseFee       *hexutil.Big
-	Hash          common.Hash `json:"hash"` // adds call to Hash() in MarshalJSON
-	BlobGasUsed   *hexutil.Uint64
-	ExcessBlobGas *hexutil.Uint64
```
Comment (jed-wemade) [inline core/types/block.go:113]: `Header` still has the fields so I recommend them restored.
And run `go generate` to output autogen files.

Header verification would be done in `VerifyHeader()`.

File: core/types/transaction.go [hunk 5]
```diff
@@ -45,10 +45,11 @@ var (
...
 const (
-	LegacyTxType     = 0x00
-	AccessListTxType = 0x01
-	DynamicFeeTxType = 0x02
-	BlobTxType       = 0x03
+	LegacyTxType                = 0x00
+	AccessListTxType            = 0x01
+	DynamicFeeTxType            = 0x02
+	BlobTxType                  = 0x03
+	FeeDelegateDynamicFeeTxType = 22 // fee delegation
```
Comment (jed-wemade) [inline core/types/transaction.go]: For consistency:
```go
	FeeDelegateDynamicFeeTxType = 0x16 // fee delegation (22)
```

File: core/types/transaction_signing.go [hunk 6]
```diff
@@ -244,6 +265,73 @@ func (s cancunSigner) Hash(tx *Transaction) common.Hash {
...
+		return nil, nil, nil, ErrInvalidChainId
+	}
+	R, S, _ = decodeSignature(sig)
+	V = big.NewInt(int64(sig[64]))
+	return R, S, V, nil
+}
+
+// Hash returns the hash to be signed by the sender.
+// It does not uniquely identify the transaction.
+func (s feeDelegateSigner) Hash(tx *Transaction) common.Hash {
```
Comment (jed-wemade) [inline core/types/transaction_signing.go:312]: Tx type check before main logic.
```go
	if tx.Type() != FeeDelegateDynamicFeeTxType {
		return s.londonSigner.Hash(tx)
	}
```

File: core/wemix_genesis.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+  "config": {
+    "berlinBlock": 0,
+    "byzantiumBlock": 0,
+    "chainId": 1111,
+    "constantinopleBlock": 0,
+    "eip150Block": 0,
+    "eip155Block": 0,
+    "eip158Block": 0,
+    "homesteadBlock": 0,
+	"petersburgBlock": 0,
```
Comment (jed-wemade) [inline core/wemix_genesis.go]: Tab -> 4 spaces. same for line 318.

File: ethclient/test/ethclient_test.go [hunk 8]
```diff
@@ -14,7 +14,9 @@
 // You should have received a copy of the GNU Lesser General Public License
 // along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
 
-package ethclient
+// this file uses `github.com/ethereum/go-ethereum/node` package which uses `github.com/ethereum/go-ethereum/ethclient` package
+// so we should put it on `test` package under `ethclient` to avoid import cycle
```
Comment (jed-wemade) [inline ethclient/test/ethclient_test.go:18]: Without changing the test code, we can resolve the import cycle.
```diff
diff --git a/node/ethclient/client.go b/node/ethclient/client.go
new file mode 100644
index 000000000..91c8baeb8
--- /dev/null
+++ b/node/ethclient/client.go
@@ -0,0 +1,10 @@
+package ethclient
+
+import (
+       "github.com/ethereum/go-ethereum/ethclient"
+       "github.com/ethereum/go-ethereum/rpc"
+)
+
+func NewClient(c *rpc.Client) *ethclient.Client {
+       return ethclient.NewClient(c)
+}
diff --git a/node/node.go b/node/node.go
index a8aec6624..52d8ef8d1 100644
--- a/node/node.go
+++ b/node/node.go
@@ -32,10 +32,10 @@ import (
        "github.com/ethereum/go-ethereum/common"
        "github.com/ethereum/go-ethereum/common/hexutil"
        "github.com/ethereum/go-ethereum/core/rawdb"
-       "github.com/ethereum/go-ethereum/ethclient"
        "github.com/ethereum/go-ethereum/ethdb"
        "github.com/ethereum/go-ethereum/event"
        "github.com/ethereum/go-ethereum/log"
+       "github.com/ethereum/go-ethereum/node/ethclient"
        "github.com/ethereum/go-ethereum/p2p"
        "github.com/ethereum/go-ethereum/rpc"
        "github.com/ethereum/go-ethereum/wemixgov"
```

File: params/config.go [hunk 9]
```diff
@@ -453,23 +580,33 @@ func (c *ChainConfig) Description() string {
 	if c.ArrowGlacierBlock != nil {
 		banner += fmt.Sprintf(" - Arrow Glacier:               #%-8v (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/arrow-glacier.md)\n", c.ArrowGlacierBlock)
 	}
+	banner += fmt.Sprintf(" - London:                      #%-8v (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/london.md)\n", c.LondonBlock)
```
Comment (jed-wemade) [inline params/config.go]: Dup in line 579

File: params/config.go [hunk 10]
```diff
@@ -426,7 +553,7 @@ func (c *ChainConfig) Description() string {
 	case c.QBFT != nil:
 		banner += "Consensus: QBFT (quorum-byzantine-fault-tolerance)\n"
 	default:
-		banner += "Consensus: unknown\n"
+		banner += "Consensus: WEMIX PoA\n"
```
Comment (jed-wemade) [inline params/config.go:584]: Default should be "unknown" to know wrong consensus setup.
```go
	case c.Brioche != nil:
		banner += "Consensus: WEMIX PoA\n"
	default:
		banner += "Consensus: unknown\n"
```

File: params/protocol_params.go [hunk 11]
```diff
@@ -189,3 +189,10 @@ var (
 	// SystemAddress is where the system-transaction is sent from as per EIP-4788
 	SystemAddress common.Address = common.HexToAddress("0xfffffffffffffffffffffffffffffffffffffffe")
 )
+
+// wemix parameters
+var (
+	FixedDifficulty           int64  = 1 // 0 means no fixed difficulty
+	FixedGasLimit             uint64 = 0 // 0 means no fixed gas limit,
```
Comment (jed-wemade) [inline params/protocol_params.go]: Unused constant `FixedGasLimit`

File: rpc/handler.go [hunk 12]
```diff
@@ -475,7 +475,7 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage) *jsonrpcMess
 			if resp.Error.Data != nil {
 				ctx = append(ctx, "errdata", resp.Error.Data)
 			}
-			h.log.Warn("Served "+msg.Method, ctx...)
+			h.log.Debug("Served "+msg.Method, ctx...)
```
Comment (jed-wemade) [inline rpc/handler.go:478]: If log is not noisy, `Warn` log level is suitable.
Are there lots of logs printed?

File: wemixgov/bind/const.go
Comment (jed-wemade) [inline wemixgov/bind/const.go:1]: To resolve ambiguity adding prefix `CONTRACT_` would be helpful. (i.e., `REGISTRY` -> `CONTRACT_REGISTRY`)
Plus, for strict type check:
```go
type GovContract string
const (
	CONTRACT_REGISTRY          = GovContract("Registry")
	...

type GovDomain string
const (
	DOMAIN_Gov           = GovDomain("GovernanceContract")
	...
```

File: core/types/block.go (hunk 4 above)
Comment (egonspace) [reply core/types/block.go:113]: We need to remove these fields for the compatibility of wemix legacy.

File: params/config.go (hunk 10 above)
Comment (egonspace) [reply params/config.go:584]: Currently, geth uses WEMIX consensus engine for default.

File: rpc/handler.go (hunk 12 above)
Comment (egonspace) [reply rpc/handler.go:478]: Yes, it was too loud. go-wemix uses `Debug` for it.

File: wemixgov/bind/const.go
Comment (egonspace) [reply wemixgov/bind/const.go:1]: That code says a compile error: `errors.Wrap(err, CONTRACT_REGISTRY)`
We must do type-cast to `string`

Comment (jed-wemade) [reply wemixgov/bind/const.go:1]: > We must do type-cast to string

This is what I intended.
We can restrict them by using type check to reduce mistake.
But on second thought, depite of the change, we can misuse the domain for current struct.

So the types is not effective at this point.

**Review by kai-yu-wm — APPROVED** [review]

**Pull request merged**